scripts/bo_benchmarks/demonstrations/branin_campaign_demonstration.py

```diff
@@ -0,0 +1,419 @@
+        print(f"Ax cross validation failed: {e}")
+        return None, None
+
+
+def calculate_iteration_metrics(ax_client, ax_client_cv, trial_index, x1, x2, result):
+    """Calculate evaluation metrics after each trial using both GP model and Ax cross validation."""
+    # Skip if not enough trials
+    if trial_index < 3:
+        return None
+    
+    # Get all completed trials data
+    df = ax_client.get_trials_data_frame()
+    if len(df) < 3:
+        return None
+    
+    try:
+        # ===== GP-based metrics (gpcheck) =====
+        # Prepare data for GP model
+        train_X = torch.tensor(df[['x1', 'x2']].values, dtype=torch.float32)
+        train_Y = torch.tensor(df[obj1_name].values, dtype=torch.float32).unsqueeze(-1)
+        
+        # Fit GP model using gpcheck
+        config = GPConfig(seed=42)
+        gp_model = GPModel(config)
+        gp_model.fit(train_X, train_Y)
+        
+        # Calculate GP-based metrics
+        mean_pred, std_pred = gp_model.predict(train_X)
+        
+        gp_r2 = r2_score(train_Y.detach().cpu().numpy(), mean_pred.detach().cpu().numpy())
+        
+        ls, imp = gp_model.get_lengthscales()
+        imp_dict = importance_concentration(imp)
+        
+        loo = loo_pseudo_likelihood(gp_model.model, train_X, train_Y)
+        
+        rank_tau = kendalltau(train_Y.squeeze().detach().cpu().numpy(), mean_pred.detach().cpu().numpy())[0]
+        
+        # Calculate interval score using GP predictions
+        y_true = train_Y.squeeze().detach().cpu().numpy()
+        y_pred_mean = mean_pred.detach().cpu().numpy()
+        y_pred_std = std_pred.detach().cpu().numpy()
+        
+        # Convert to 95% confidence intervals
+        lower, upper = normal_95ci(y_pred_mean, y_pred_std)
+        interval_scores = interval_score(y_true, lower, upper)
+        mean_interval_score = np.mean(interval_scores)
+        
+        # ===== Ax cross validation metrics =====
+        ax_cv_r2, ax_cv_interval_score = calculate_ax_cross_validation_metrics(ax_client_cv, trial_index, x1, x2, result)
+        
+        metrics = {
+            'trial': trial_index,
+            'gp_r2': gp_r2,  # GP-based R²
+            'ax_cv_r2': ax_cv_r2,  # Ax cross validation R²
+            'interval_score': mean_interval_score,  # Mean interval score (GP-based)
+            'ax_cv_interval_score': ax_cv_interval_score,  # Mean interval score (Ax CV)
+            'imp_cumsum': imp_dict,
+            'loo_nll': loo,
+            'rank_tau': rank_tau,
+            'lengthscales': ls,
+            'importance': imp
+        }
+        
+        # Print metrics for this iteration
+        print(f"  GP R²: {gp_r2:.4f}, LOO NLL: {loo:.2f}")
+        print(f"  GP Interval Score: {mean_interval_score:.4f}")
+        
+        return metrics
+        
+    except Exception as e:
+        print(f"Metrics calculation failed: {e}")
+        return None
+
+
+# Create main ax client for optimization
+ax_client = AxClient()
```

Comment:
This should be a custom generation strategy where it goes from Sobol to BOTORCH_MODULAR, except with Sobol fixed to 4 trials.